src/ReactFinalForm.js

```diff
@@ -34,63 +36,128 @@ export const all: FormSubscription = formSubscriptionItems.reduce(
   },
   {}
 )
+const validationStateChanged = (a: FormState, b: FormState) =>
+  ['valid', 'invalid', 'errors'].some(key => !shallowEqual(a[key], b[key]))
 
-type State = {
-  state: FormState
-}
+const ReactFinalForm = ({
+  debug,
+  decorators,
+  destroyOnUnregister,
+  initialValues,
+  initialValuesEqual,
+  keepDirtyOnReinitialize,
+  mutators,
+  onSubmit,
+  subscription,
+  validate,
+  validateOnBlur,
+  ...rest
+}: Props) => {
+  const config: Config = {
+    debug,
+    destroyOnUnregister,
+    initialValues,
+    keepDirtyOnReinitialize,
+    mutators,
+    onSubmit,
+    validate,
+    validateOnBlur
+  }
 
-class ReactFinalForm extends React.Component<Props, State> {
-  props: Props
-  state: State
-  form: FormApi
-  mounted: boolean
-  resumeValidation: ?boolean
-  unsubscriptions: Unsubscribe[]
+  const form: FormApi = useConstant(() => {
+    const f = createForm(config)
+    f.pauseValidation()
+    return f
+  })
+  const firstRender = React.useRef(true)
 
-  constructor(props: Props) {
-    super(props)
-    const {
-      children,
-      component,
-      render,
-      subscription,
-      decorators,
-      ...rest
-    } = props
-    const config: Config = rest
-    this.mounted = false
-    try {
-      this.form = createForm(config)
-    } catch (e) {
-      // istanbul ignore next
-      if (process.env.NODE_ENV !== 'production') {
-        console.error(`Warning: ${e.message}`)
-      }
-    }
-    this.unsubscriptions = []
-    if (this.form) {
-      // set initial state
+  // synchronously register and unregister to query form state for our subscription on first render
+  const [state, setState] = React.useState<FormState>(
+    (): FormState => {
       let initialState: FormState = {}
-      this.form.subscribe((state: FormState) => {
+      form.subscribe(state => {
```

Comment:
maybe this is an opportunity for a getState on the form? or do you want to discourage using it as part of the public API?

Reply:
It _could_ be, but I think it should respect the subscription. It shouldn't give you all of the form state on first render and then to have it restricted to your subscription on subsequent renders.

Great thought, though.

Reply:
Oh, thought subscription specifies which keys should trigger a callback, didnt know it also specifies which exact keys you are going to receive. Makes sense that way then.